File: bridge/zig/src/pty.zig
const std = @import("std");
const posix = std.posix;
const c = @cImport({
    @cInclude("pty.h");
    @cInclude("sys/ioctl.h");
    @cInclude("sys/wait.h");
});

pub const Pty = struct {
    master_fd: posix.fd_t,
    child_pid: posix.pid_t,
    alive: bool = true,

    pub fn spawn(shell: [*:0]const u8) !Pty {
        var master_fd: c_int = undefined;
        const pid = c.forkpty(&master_fd, null, null, null);
        if (pid < 0) return error.ForkFailed;

        if (pid == 0) {
            const env = [_:null]?[*:0]const u8{ "TERM=xterm-256color", "PATH=/usr/bin:/bin", null };
            const argv = [_:null]?[*:0]const u8{ shell, null };
            _ = posix.execvpeZ(shell, &argv, &env) catch {};
            posix.exit(1);
        }

        return .{ .master_fd = master_fd, .child_pid = @intCast(pid) };
    }

    pub fn resize(self: Pty, rows: u16, cols: u16) void {
        var ws = c.winsize{ .ws_row = rows, .ws_col = cols, .ws_xpixel = 0, .ws_ypixel = 0 };
        _ = c.ioctl(self.master_fd, c.TIOCSWINSZ, &ws);
    }

    pub fn writeAll(self: Pty, data: []const u8) !void {
        var written: usize = 0;
        while (written < data.len) {
            written += posix.write(self.master_fd, data[written..]) catch |err| switch (err) {
                error.WouldBlock => continue,
                else => return err,
            };
        }
    }

    pub fn read(self: Pty, buf: []u8) !usize {
        return posix.read(self.master_fd, buf) catch |err| switch (err) {
            error.InputOutput => return 0, // EIO = PTY closed
            else => return err,
        };
    }

    pub fn signal(self: Pty, sig: u8) bool {
        // Whitelist safe signals
        const allowed = [_]u8{ 2, 3, 9, 15, 18, 19, 20, 28 }; // INT, QUIT, KILL, TERM, CONT, STOP, TSTP, WINCH
        for (allowed) |s| {
            if (sig == s) {
                posix.kill(self.child_pid, sig) catch return false;
                return true;
            }
        }
        return false;
    }

    pub fn reap(self: *Pty) ?i32 {
        if (!self.alive) return null;
        var status: c_int = 0;
        const ret = c.waitpid(self.child_pid, &status, c.WNOHANG);
        if (ret > 0) {
            self.alive = false;
            if (c.WIFEXITED(status)) return @intCast(c.WEXITSTATUS(status));
            if (c.WIFSIGNALED(status)) return -@as(i32, @intCast(c.WTERMSIG(status)));
        }
        return null;
    }

    pub fn close(self: *Pty) ?i32 {
        posix.close(self.master_fd);
        // Reap child, blocking if needed
        if (self.alive) {
            var status: c_int = 0;
            _ = c.waitpid(self.child_pid, &status, 0);
            self.alive = false;
            if (c.WIFEXITED(status)) return @intCast(c.WEXITSTATUS(status));
            if (c.WIFSIGNALED(status)) return -@as(i32, @intCast(c.WTERMSIG(status)));
        }
        return null;
    }

    pub fn fd(self: Pty) posix.fd_t {
        return self.master_fd;
    }
};

File: bridge/c/pty.c
#define _POSIX_C_SOURCE 200809L
#define _DEFAULT_SOURCE
#include "pty.h"

#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/wait.h>

#if defined(__APPLE__) || defined(__FreeBSD__)
#  include <util.h>
#else
#  include <pty.h>
#endif

int bridge_pty_spawn(bridge_pty_t *p, const char *shell) {
    int master_fd = -1;
    pid_t pid = forkpty(&master_fd, NULL, NULL, NULL);
    if (pid < 0) return -1;

    if (pid == 0) {
        // Child
        char *env[] = {
            (char *)"TERM=xterm-256color",
            (char *)"PATH=/usr/bin:/bin",
            NULL,
        };
        char *argv[] = { (char *)shell, NULL };
        execve(shell, argv, env);
        _exit(1);
    }

    p->master_fd = master_fd;
    p->child_pid = pid;
    p->alive = 1;
    return 0;
}

void bridge_pty_resize(bridge_pty_t *p, uint16_t rows, uint16_t cols) {
    struct winsize ws = { .ws_row = rows, .ws_col = cols, 0, 0 };
    (void)ioctl(p->master_fd, TIOCSWINSZ, &ws);
}

int bridge_pty_write_all(bridge_pty_t *p, const void *buf, size_t len) {
    const uint8_t *b = buf;
    size_t written = 0;
    while (written < len) {
        ssize_t n = write(p->master_fd, b + written, len - written);
        if (n < 0) {
            if (errno == EINTR || errno == EAGAIN) continue;
            return -1;
        }
        written += (size_t)n;
    }
    return 0;
}

long bridge_pty_read(bridge_pty_t *p, void *buf, size_t len) {
    ssize_t n = read(p->master_fd, buf, len);
    if (n < 0) {
        // EIO on Linux when the slave side closes — treat as EOF.
        if (errno == EIO) return 0;
        return -1;
    }
    return (long)n;
}

int bridge_pty_signal(bridge_pty_t *p, int sig) {
    static const int allowed[] = {
        SIGINT, SIGQUIT, SIGKILL, SIGTERM, SIGCONT, SIGSTOP, SIGTSTP, SIGWINCH
    };
    for (size_t i = 0; i < sizeof(allowed)/sizeof(allowed[0]); i++) {
        if (sig == allowed[i]) {
            return kill(p->child_pid, sig) == 0 ? 1 : 0;
        }
    }
    return 0;
}

static int decode_status(int status) {
    if (WIFEXITED(status)) return WEXITSTATUS(status);
    if (WIFSIGNALED(status)) return -WTERMSIG(status);
    return 0;
}

int bridge_pty_reap(bridge_pty_t *p, int *code) {
    if (!p->alive) return 0;
    int status = 0;
    pid_t ret = waitpid(p->child_pid, &status, WNOHANG);
    if (ret > 0) {
        p->alive = 0;
        *code = decode_status(status);
        return 1;
    }
    return 0;
}

int bridge_pty_close(bridge_pty_t *p) {
    if (p->master_fd >= 0) {
        close(p->master_fd);
        p->master_fd = -1;
    }
    int code = 0;
    if (p->alive) {
        int status = 0;
        waitpid(p->child_pid, &status, 0);
        p->alive = 0;
        code = decode_status(status);
    }
    return code;
}
